Serialization improvements

sipa opened the pull request:
This PR improves correctness (removing potentially unsafe `const_cast`s) and flexibility of the serialization code.

The main issue is that use of the current `ADD_SERIALIZE_METHODS` macro (which is the only way to not duplicate serialization and deserialization code) only expands to a single class method, and thus can only be qualified as either const or non-const - not both. In many cases, serialization needs to work on const objects however, and preferably that is done without casts that could hide const-correctness bugs.

To deal with that, this PR introduces a new approach that includes a `SERIALIZE_METHODS(obj)` macro, where `obj` is a variable name. It expands to some boilerplate and a static method to which the object itself is an argument. The advantage is that its type can be templated, and be `const` when serializing.

Another issue is the various serialization-wrapping macros (`VARINT`, `COMPACTSIZE`, `FLATDATA` and `LIMITED_STRING`). They all `const_cast` their argument in order to construct a wrapper object, which supports both serialization and deserialization. This PR makes them templated in the underlying data type (for example, `CompactSizeWrapper<uint64_t>`). This has the advantage that we can make the template type `const` when invoked on a `const` variable (so it would be `CompactSizeWrapper<const uint64_t>` in that case).

A last issue is the persistent use of the `REF` macro to deal with temporary expressions being passed in. Since C++11, this is not needed anymore as temporaries are explicitly represented as rvalue references. Thus we can remove `REF` invocations and instead just make the various classes and helper functions deal correctly with references.

The above changes permit a fully const-correct version of all serialization code. However, it is cumbersome. Any existing `ADD_SERIALIZE_METHODS` instances in the code that do more than just (conditionally) serializing/deserializing some fields (in particular, it contains branches that assign to some of the variables) need to be split up into an explicit `Serialize` and `Unserialize` method instead. In some cases this is inevitable (wallet serializers do some crazy transformations while serializing!), but in many cases it is just annoying duplication.

To improve upon this, a few more primitives that are currently inlined are turned into serialization wrappers:
* `BigEndianWrapper`: Serializes/deserializes an integer as big endian rather than little endian (only for 16-bit). This permits the CService serialization to become a oneliner.
* `Uint48Wrapper`: Serializes/deserializes only the lower 48 bits of an integer (used in BIP152 code).
* `VectorApplyWrapper`: Serializes/deserializes a vector while using a custom serializer for its elements. This simplifies the undo and blockencoding serializers a lot.

Best of all, it removes 147 lines of while code adding a bunch of comments (though the increased use of vararg `READWRITE` is probably cheating a bit).

The commits are ordered into 3 sections:
* First, introduce new classes that permit const-correct serialization.
* Then one by one transform the various files to use the new serializers.
* Finally, remove the old serializers.

This may be too much to go in at once. I'm happy to split things up as needed.

sipa committed:
Introduce new serialization macros without casts

This new approach uses a static method which takes the object as
a argument. This has the advantage that its constness can be a
template parameter, allowing a single implementation that sees the
object as const for serialization and non-const for deserialization,
without casts.

More boilerplate is included in the new macro as well.

sipa committed:
Convert primitives to new serialization

sipa committed:
Convert addrdb/addrman to new serialization

sipa committed:
Generalize CompactSize wrapper

This makes it const-correct and usable for other integer types.

sipa committed:
Generalize VarInt wrappers

sipa committed:
Generalize LimitedString wrapper

sipa committed:
Convert blockencodings to new serialization

sipa committed:
Convert merkleblock/bloom to new serialization

sipa committed:
Convert chain to new serialization

sipa committed:
Convert netaddress to new serialization

sipa committed:
Generalize BigEndian serialization wrapper

sipa committed:
Convert feerate to new serialization

sipa committed:
Convert protocol/net to new serialization

sipa committed:
Convert compressor/txdb/coins/undo/script to new serialization

sipa committed:
Convert Qt to new serialization

sipa committed:
Convert rest to new serialization

sipa committed:
Convert dbwrapper tests to new serialization

sipa committed:
Convert serialize_tests/streams_tests to new serialization

sipa committed:
Convert wallet/walletdb/crypter to new serialization

sipa committed:
Remove old serialization primitives

sipa committed:
Add custom vector-element serialization wrapper

This allows a very compact notation for serialization of vectors whose
elements are not serialized using their default encoding.

gmaxwell commented:
Should probably be tested on big endian. :)

sipa commented on src/blockencodings.h:
```diff
@@ -12,58 +12,70 @@
+    void Unserialize(Stream& s)
+    {
+        std::vector<uint64_t> tmp;
+        s >> blockhash >> VectorApply<CompactSizeWrapper>(tmp);
```
For ease of implementation, deserialization first happens into a uint64_t, and is then converted. This means a temporary is created and allocated, which is an overhead that the old implementation didn't have.

If this is considered too much, an alternative is to lose the overflow checking, and instead use `s >> blockhash >> VectorApply<CompactSizeWrapper>(indexes)` directly, which will work, but compute results modulo 2^16.

sipa commented on src/test/serialize_tests.cpp:
```diff
@@ -49,11 +48,10 @@ class CSerializeMethodsTestMany : public CSerializeMethodsTestSingle
-        READWRITEMANY(intval, boolval, stringval, FLATDATA(charstrval), txval);
+    SERIALIZE_METHODS(obj)
+    {
+        READWRITE(obj.intval, obj.boolval, obj.stringval, FlatData(obj.charstrval), obj.txval);
```
This whole test is somewhat less valuable now, as both cases use `READWRITE`.

sipa commented on src/wallet/wallet.h:
```diff
@@ -378,42 +376,36 @@ class CWalletTx : public CMerkleTx
-    inline void SerializationOp(Stream& s, Operation ser_action) {
-        if (ser_action.ForRead())
-            Init(NULL);
+    template<typename Stream>
```
This is one of the more involved changes, as it's both splitting the serializer into two versions, and the `Serialize` code no longer modifies `mapValue` in-place (wtf?).

sipa commented on src/wallet/wallet.h:
```diff
@@ -378,42 +376,36 @@ class CWalletTx : public CMerkleTx
+    }
 
-            ReadOrderPos(nOrderPos, mapValue);
+    template<typename Stream>
```
Here is another big change, that avoids modifying `mapValue` and `strAccount` (and then later fixing it up before returning!).

sipa commented on src/serialize.h:
```diff
@@ -456,14 +458,104 @@ class LimitedString
+ * V is not required to be an std::vector type. It works for any class that
+ * exposes a value_type, iteration, and resize method that behave like vectors.
+ */
+template<template <typename> class W, typename V> class VectorApplyWrapper
```
Notice the unusual construction of a template that takes a template as parameter here. See "Template template parameter" here: http://en.cppreference.com/w/cpp/language/template_parameters

sipa commented on src/serialize.h:
```diff
@@ -898,16 +978,16 @@ void SerializeMany(Stream& s)
+void SerializeMany(Stream& s, const Arg& arg)
 {
-    ::Serialize(s, std::forward<Arg>(arg));
+    ::Serialize(s, arg);
```
The reason for removing the `std::forward` calls here is explained in the commit message (there is no benefit in passing down the rvalue-ness).

sipa reviewed (commented):
Here is a self-review in which I point out some of the things reviewers may want to be aware of.

jonasschnelli commented:
Concept ACK.
Binaries: https://bitcoin.jonasschnelli.ch/build/210 (Currently running on a fresh node)
Agree with @gmaxwell that some BE testing would be good.

Will code-review soon.

sipa commented:
Made some changes to reduce the size of the overall diff.

promag commented on src/serialize.h:
```diff
@@ -495,11 +494,14 @@ class LimitedString
 };
+//! Add a LimitedString wrapper around a const string (identity)
+template<size_t I> static inline const std::string& LimitedString(const std::string& str) { return str; }
+//! Add a LimitedString wrapper around a non-const string
```
Missing periods in comments.

promag reviewed (commented):
ACK b0652ac. Everything works as expected.

Indeed a lot in one shot :).

sipa commented:
Rebased.

sipa commented:
Rebased.

ryanofsky commented on src/serialize.h:
```diff
@@ -446,6 +428,16 @@ class CFlatData
+//! Construct a FlatRange wrapper around a non-const POD and array types.
+template<typename T> static inline FlatRangeWrapper<char> FlatDataInner(T* t, size_t len) { return FlatRangeWrapper<char>((char*)t, ((char*)t) + len); }
+//! Helper macro to easily serialize POD types.
+#define FLATDATA(x) FlatDataInner(&(x), sizeof(x))
```
In commit "Generalize FlatData wrapper"

Though these changes don't make FLATDATA more dangerous than it was previously, the lack of type safety relying on C casts and sizeof here is a little scary. I experimented a little, and it seems this could be cleaned up with some simple changes I posted here: 6ef78bcd83dd6f88362dec29736811b762ad75eb. Could you take a look, and maybe incorporate these into the PR?

sipa replied:
I've incorporated part of your changes, but gone further and just added native support for serializing char arrays (without any wrapper). I think this is much cleaner now.

ryanofsky reviewed (commented):
Reviewed first several commits. Will continue reviewing.

sipa commented on src/serialize.h:
```diff
@@ -468,23 +467,27 @@ class CVarInt
-    void Serialize(Stream &s) const {
-        WriteCompactSize<Stream>(s, n);
+    void Unserialize(Stream& s) {
+        n = ReadCompactSize<Stream>(s);
```
Fixed.

sipa commented on src/serialize.h:
```diff
@@ -468,23 +467,27 @@ class CVarInt
-    void Unserialize(Stream& s) {
-        n = ReadCompactSize<Stream>(s);
+    void Serialize(Stream& s) const {
+        WriteCompactSize<Stream>(s, n);
```
Fixed.

sipa commented on src/serialize.h:
```diff
@@ -493,9 +515,44 @@ class LimitedString
 
+/** Serialization wrapper class for big-endian integers.
+ *
+ * Use this wrapper around integer types that are stored in memory in native
```
Fixed.

sipa commented on src/serialize.h:
```diff
@@ -493,9 +515,44 @@ class LimitedString
+ * Use this wrapper around integer types that are stored in memory in native
+ * byte order, but serialized in big endian notation.
+ *
+ * Onlyy 16-bit types are supported for now.
```
Fixed.

sipa commented on src/serialize.h:
```diff
@@ -493,9 +515,44 @@ class LimitedString
+public:
+    explicit BigEndianWrapper(I& val) : m_val(val)
+    {
+        static_assert(S == 2 || S == 4, "Unsupported BigEndian size");
```
Fixed.

ryanofsky commented on src/netaddress.h:
```diff
@@ -132,7 +132,7 @@ class CSubNet
             READWRITE(network);
             READWRITE(FLATDATA(netmask));
-            READWRITE(FLATDATA(valid));
+            READWRITE(valid);
```
In commit "Overhaul FLATDATA"

Might be worth splitting this change out into separate commit, or noting in commit message here that that this change is not backwards compatible on platforms where sizeof(bool) is not 1.

sipa replied:
Done.

ryanofsky commented on src/serialize.h:
```diff
@@ -406,48 +415,36 @@ I ReadVarInt(Stream& is)
- * Wrapper for serializing arrays and POD.
- */
-class CFlatData
+/** Wrapper for serializing arrays and POD. */
```
In commit "Overhaul FLATDATA"

Probably more accurate to say "wrapper for serializing char arrays". (Though in principle this could work with stream classes with read/write methods not taking char pointers.)

sipa replied:
Fixed.

ryanofsky commented on src/serialize.h:
```diff
@@ -406,48 +415,36 @@ I ReadVarInt(Stream& is)
     }
 };
+//! Construct a FlatRange wrapper around a const char vector.
+template<typename T> static inline const FlatRangeWrapper<const char> FlatVector(const T& t) { return FlatRangeWrapper<const char>(CharCast(t.data()), CharCast(t.data() + t.size())); }
```
In commit "Overhaul FLATDATA"

Maybe call it CharVector or CharArray instead of FlatVector. FlatVector is kind of redundant because any vectors should be flat. But also the vector part is limiting because these functions can work for other types (std::array, std::string, SecureString, std::basic_string_view, etc)

sipa replied:
Done.

sipa commented on src/serialize.h:
```diff
@@ -470,23 +469,36 @@ class CVarInt
-        n = ReadCompactSize<Stream>(s);
+    void Serialize(Stream& s) const
+    {
+        WriteCompactSize<Stream>(s, m_n);
```
Done.

sipa commented on src/serialize.h:
```diff
@@ -495,9 +517,48 @@ class LimitedString
+    template<typename Stream>
+    void Unserialize(Stream& s)
+    {
+        if (S == 2) m_val = ser_readdata16be(s);
```
Done.

ryanofsky commented on src/serialize.h:
```diff
@@ -573,6 +573,64 @@ template<typename I> class BigEndianWrapper
+            }
+            m_vector.resize(nMid);
+            for (value_type& x : m_vector) {
+                s >> W<value_type>(x);
```
In commit "Add custom vector-element serialization wrapper"

Doesn't this overwrite elements in the front of the vector each time through the while loop?

sipa replied:
Yes, this was bogus. Thanks for pointing that out; fixed by rewriting,

ryanofsky commented on src/blockencodings.h:
```diff
@@ -12,58 +12,70 @@
     }
 };
 
+template<typename T>
```
In commit "Convert blockencodings to new serialization"

Maybe move this up closer to TransactionCompressWrapper class

sipa replied:
Done.

ryanofsky commented on src/blockencodings.h:
```diff
@@ -154,40 +156,18 @@ class CBlockHeaderAndShortTxIDs {
+    {
+        static_assert(SHORTTXIDS_LENGTH == 6, "shorttxids serialization assumes 6-byte shorttxids");
+        s >> header >> nonce >> VectorApply<Uint48Wrapper>(shorttxids) >> prefilledtxn;
+        FillShortTxIDSelector();
```
In commit "Convert blockencodings to new serialization"

Would you be opposed to adding mutable object access to SERIALIZE_METHODS so Serialize and Unserialize methods don't need to be split up? I could think of a number of ways to do this. Maybe easiest would be to stick a mutable pointer inside ser_action:

```
SERIALIZE_METHODS(obj) {
  READWRITE(obj.header, obj.nonce, VectorApply<Uint48Wrapper>(obj.shorttxids), obj.prefilledtxn);
  if (ser_action.ForRead()) ser_action.MutableObj()->FillShortTxId();
}
```

The pointer would be null when serializing.

sipa replied:
That's a neat trick. It seems a bit ugly to need to fake returning a mutable object nullptr. I have an alternative, but I'm not sure it's any better.:

```c++
template <typename T, typename F>
void IfUnserializer(T& obj, CSerActionUnserialize ser_action, const F& fn) { fn(obj); }
template <typename T, typename F>
void IfUnserialize(const T& obj, CSerActionSerialize ser_action, const F& fn) {}
#define IF_UNSERIALIZE(typ, obj, code) (::IfUnserialize<typ>(obj, ser_action, [&](typ& obj)code))
```

Which you'd then invoke using

```c++
     SERIALIZE_METHODS(SomeType, obj)
     {
          READWRITE(obj.member);
          IF_UNSERIALIZE(SomeType, obj, {obj.Cleanup();});
     }
```

sipa replied:
@ryanofsky I don't see how to get rid of passing in the type to IF_UNSERIALIZE.

ryanofsky replied:
@ryanofsky I don't see how to get rid of passing in the type to IF_UNSERIALIZE.

You could pass it as a template parameter to CSerActionSerialize / CSerActionUnserialize and access it through ser_action.

ryanofsky reviewed (commented):
More review comments. I made it up to the "Convert blockencodings to new serialization" commit this time.

sipa commented on src/serialize.h:
```diff
@@ -148,9 +148,21 @@ enum
     SER_GETHASH         = (1 << 2),
 };
 
+// Convert the reference base type to X, without changing constness or reference type.
```
Done.

sipa commented on src/serialize.h:
```diff
@@ -167,6 +167,28 @@ enum
+ * thus allows a single implementation that sees the object as const for serializing
+ * and non-const for deserializing, without casts.
+ */
+#define SERIALIZE_METHODS(obj)                                                      \
```
Done!

sipa commented on src/serialize.h:
```diff
@@ -898,16 +898,16 @@ void SerializeMany(Stream& s)
 
 template<typename Stream, typename Arg>
-void SerializeMany(Stream& s, Arg&& arg)
+void SerializeMany(Stream& s, const Arg& arg)
```
Done.

sipa commented on src/serialize.h:
```diff
@@ -519,19 +518,23 @@ class LimitedString
-            s.read((char*)string.data(), size);
+        m_string.resize(size);
+        if (size != 0) {
+            s.read((char*)m_string.data(), size);
```
Done.

sipa commented on src/serialize.h:
```diff
@@ -450,24 +449,29 @@ class CFlatData
     }
 };
+//! Automatically construct a VarInt wrapper around the argument.
+template<typename I> static inline VarIntWrapper<typename std::remove_reference<I>::type> VARINT(I&& i) { return VarIntWrapper<typename std::remove_reference<I>::type>(i); }
```
The problem is that VARINT is called with temporaries as arguments, which is not true for the other ones. Either it's written as 4 cases, or using `std::remove_refence`.

sipa replied:
Oh, you're right. An lvalue reference parameter binds to rvalue reference argument, so all good.

sipa replied:
Done.

sipa commented on src/serialize.h:
```diff
@@ -189,22 +170,6 @@ template<typename X> const X&& AsBaseType(const X&& x) { return std::move(x); }
- * code. Adding "ADD_SERIALIZE_METHODS" in the body of the class causes these wrappers to be
- * added as members. 
- */
-#define ADD_SERIALIZE_METHODS                                         \
```
Done.

sipa commented on src/serialize.h:
```diff
@@ -573,6 +573,64 @@ template<typename I> class BigEndianWrapper
+        unsigned int nSize = ReadCompactSize(s);
+        unsigned int nMid = 0;
+        while (nMid < nSize) {
+            nMid += 5000000 / sizeof(value_type);
```
Done. The code was also totally broken, so I've rewritten it.

sipa commented on src/serialize.h:
```diff
@@ -470,23 +469,37 @@ class CVarInt
     }
 };
+//! Automatically construct a CompactSize wrapper around the argument.
+template<typename I> static inline CompactSizeWrapper<I> COMPACTSIZE(I& i) { return CompactSizeWrapper<I>(i); }
```
Done (and for other wrappers).

sipa commented on src/serialize.h:
```diff
@@ -406,48 +415,38 @@ I ReadVarInt(Stream& is)
     {
-        pbegin = (char*)v.data();
-        pend = (char*)(v.data() + v.size());
+        static_assert(sizeof(C) == 1, "CharArrayWrapper only works for char types");
```
I'm not sure that wchar_t has a well-defined in-memory representation.

sipa replied:
I think I'm talking about something else.

If you serialize a 16-bit wchar_t on a big endian system using this class, the characters will be serialized using 2 big endian bytes. If you then deserialize it on a little endian system, you won't get the same wchar_t values back.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -381,42 +379,36 @@ class CWalletTx : public CMerkleTx
+        std::vector<CMerkleTx> vUnused; //!< Used to be vtxPrev
+        s >> vUnused >> mapValue >> vOrderForm >> fTimeReceivedIsTxTime >> nTimeReceived >> fFromMe >> fSpent;
+
+        strFromAccount = mapValue["fromaccount"];
```
Could use std::move here.

sipa replied:
Done.

ryanofsky reviewed (commented):
utACK b49e84ea6dd7e2c0aacb8546c8d8c2d99d1d8214. Finally reviewed the full change.

sipa committed:
Add a constant for the maximum vector allocation (5 Mbyte)

sipa commented on src/serialize.h:
```diff
@@ -577,6 +577,66 @@ template<typename I> class BigEndianWrapper
+            // For DoS prevention, do not blindly allocate as much as the stream claims to contain.
+            // Instead, allocate in 5MiB batches, so that an attacker actually needs to provide
+            // X MiB of data to make us allocate X+5 Mib.
+            allocated = std::min(size, allocated + 5000000 / sizeof(value_type));
```
Done.

sipa commented on src/blockencodings.h:
```diff
@@ -12,17 +12,35 @@
+    {
+        uint32_t lsb = m_int & 0xffffffff;
+        uint16_t msb = (m_int >> 32) & 0xffff;
+        s << lsb << msb;
```
Added static assert; I'd like to avoid runtime impact.

sipa commented on src/blockencodings.h:
```diff
@@ -12,17 +12,35 @@
+        uint32_t lsb;
+        uint16_t msb;
+        s >> lsb >> msb;
+        m_int = (uint64_t(msb) << 32) | uint64_t(lsb);
```
Added static assert.

sipa commented on src/blockencodings.h:
```diff
@@ -32,38 +50,31 @@ class BlockTransactionsRequest {
-            }
+        s << COMPACTSIZE(indexes_size);
+        for (size_t i = 0; i < indexes.size(); i++) {
+            const uint64_t index = indexes[i] - (i == 0 ? 0 : (indexes[i - 1] + 1));
```
I'd rather avoid runtime overhead.

ryanofsky replied:
In that case, could add a comment like "this code will produce an invalid serialization if indices are not increasing."

ryanofsky reviewed (commented):
utACK a9770fbe6ef6d7d9bd5b61e4c51531af7cc2a1fc. Still looks good. Only minor changes since previous review, described in comments above.

laanwj commented:
Big concept ACK, happy to get rid of FLATDATA and similar ugly macros. This is a lot to review/test though, and reasonably high-risk.

sipa commented on src/undo.h:
```diff
@@ -18,81 +18,49 @@
-        // TODO: avoid reimplementing vector deserializer
-        uint64_t count = 0;
-        ::Unserialize(s, COMPACTSIZE(count));
-        if (count > MAX_INPUTS_PER_BLOCK) {
```
It could use slightly more memory when deserializing an otherwise invalid object, but that shouldn't change behaviour otherwise - if the number of transaction undo objects doesn't match the number of transaction in the block, it's invalid anyway.

sipa commented on src/qt/recentrequeststablemodel.h:
```diff
@@ -24,19 +24,19 @@ class RecentRequestEntry
+    {
+        unsigned int nDate;
+        s >> this->nVersion >> id >> nDate >> recipient;
+        date = QDateTime::fromTime_t(nDate);
```
Introduced a wrapper for QDateTime instead.

sipa commented on src/qt/walletmodel.h:
```diff
@@ -62,35 +62,32 @@ class SendCoinsRecipient
+        if (!sPaymentRequest.empty()) {
+            paymentRequest.parse(QByteArray::fromRawData(sPaymentRequest.data(), sPaymentRequest.size()));
         }
+        authenticatedMerchant = QString::fromStdString(sAuthenticatedMerchant);
```
Done using wrappers for QString and proto.

sipa commented:
Rebased.

ryanofsky reviewed (commented):
utACK c3e505c8551a29d12edcebbed8a69f1e798823ef. Changes since previous review were rebase and addressing a few comments. Main one being new guiutil.h wrappers for qt strings, dates, and protobufs.

MarcoFalke commented:
Travis can not run the qt tests. Are they passing for you locally?

ryanofsky commented:
They passed for me locally. Current travis failure appears to be "/home/travis/.travis/job_stages: fork: Cannot allocate memory"

MarcoFalke commented:
I guess they time out and you could try setting the `QTEST_FUNCTION_TIMEOUT` env var to 600,000. Though, it might be worthwhile to check why they take longer than before.

ryanofsky commented:
Actually they are failing for me locally. Stack trace shows an infinite recursion and looks like:

```
(gdb) bt -50
#2355058 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355059 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355060 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355061 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355062 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355063 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355064 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355065 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355066 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355067 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355068 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355069 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355070 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355071 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355072 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355073 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355074 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355075 0x00000001000dbe89 in GUIUtil::AsStdString<QString const>(QString const&) (qstring=...) at qt/guiutil.h:285
#2355076 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (qstring=...) at qt/guiutil.h:285
#2355077 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (qstring=...) at qt/guiutil.h:285
#2355078 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (ser_action=..., s=..., obj=...) at qt/walletmodel.h:69
#2355079 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (s=..., this=0x7fffffffb728) at qt/walletmodel.h:66
#2355080 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (a=..., os=...) at ./serialize.h:677
#2355081 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (arg=..., s=...) at ./serialize.h:1030
#2355082 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (arg=..., s=...) at ./serialize.h:1031
#2355083 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (arg=@0x7fffffffb718: 1, s=...) at ./serialize.h:1031
#2355084 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (arg=@0x7fffffffb710: 1, s=...) at ./serialize.h:1031
#2355085 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (ser_action=..., s=...) at ./serialize.h:1049
#2355086 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (s=..., obj=..., ser_action=...) at qt/recentrequeststablemodel.h:30
#2355087 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (s=..., this=0x7fffffffb710) at qt/recentrequeststablemodel.h:28
#2355088 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (a=..., os=...) at ./serialize.h:677
#2355089 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (obj=..., this=0x7fffffffb600) at ./streams.h:399
#2355090 0x00000001000da01d in RecentRequestsTableModel::addNewRequest(SendCoinsRecipient const&) (this=0x1014511d0, recipient=...) at qt/recentrequeststablemodel.cpp:173
#2355091 0x00000001000d2a92 in ReceiveCoinsDialog::on_receiveButton_clicked() (this=0x7fffffffbf50) at qt/receivecoinsdialog.cpp:167
#2355092 0x000000010010dd95 in ReceiveCoinsDialog::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_o=<optimized out>, _c=<optimized out>, _id=<optimized out>, _a=<optimized out>)
    at qt/moc_receivecoinsdialog.cpp:123
#2355093 0x000000010010e0e5 in ReceiveCoinsDialog::qt_metacall(QMetaObject::Call, int, void**) (this=0x7fffffffbf50, _c=QMetaObject::InvokeMetaMethod, _id=3, _a=0x7fffffffbbb0) at qt/moc_receivecoinsdialog.cpp:177
#2355094 0x00007ffff62b2ee0 in QMetaObject::activate(QObject*, int, int, void**) (sender=sender@entry=0x10163e380, signalOffset=<optimized out>, local_signal_index=local_signal_index@entry=2, argv=argv@entry=0x7fffffffbbb0) at kernel/qobject.cpp:3728
#2355095 0x00007ffff62b3537 in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) (sender=sender@entry=0x10163e380, m=m@entry=0x7ffff7098a40 <QAbstractButton::staticMetaObject>, local_signal_index=local_signal_index@entry=2, argv=argv@entry=0x7fffffffbbb0) at kernel/qobject.cpp:3578
#2355096 0x00007ffff6f112b2 in QAbstractButton::clicked(bool) (this=this@entry=0x10163e380, _t1=false) at .moc/moc_qabstractbutton.cpp:303
#2355097 0x00007ffff6c73f44 in QAbstractButtonPrivate::emitClicked() (this=0x1016cc1d0) at widgets/qabstractbutton.cpp:534
#2355098 0x00007ffff6c745e1 in QAbstractButton::click() (this=0x10163e380) at widgets/qabstractbutton.cpp:992
#2355099 0x0000000100086686 in (anonymous namespace)::TestGUI() () at qt/test/wallettests.cpp:222
#2355100 0x00007ffff628f84a in QMetaMethod::invoke(QObject*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument) const (this=this@entry=0x7fffffffce98, object=object@entry=0x7fffffffd830, connectionType=Qt::DirectConnection, 
    connectionType@entry=4294954784, returnValue=..., val0=..., val1=..., val2=..., val3=..., val4=..., val5=..., val6=..., val7=..., val8=..., val9=...) at kernel/qmetaobject.cpp:2212
#2355101 0x00007ffff6294f0d in QMetaObject::invokeMethod(QObject*, char const*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument) (obj=0x7fffffffd830, member=member@entry=0x100ca1990 "walletTests", type=4294954784, 
    type@entry=Qt::DirectConnection, ret=..., val0=..., val1=..., val2=..., val3=..., val4=..., val5=..., val6=..., val7=..., val8=..., val9=...) at kernel/qmetaobject.cpp:1479
#2355102 0x00007ffff7fa408c in QTest::qInvokeTestMethod(char const*, char const*) (val9=..., val8=..., val7=..., val6=..., val5=..., val4=..., val3=..., val2=..., val1=..., val0=..., type=Qt::DirectConnection, member=0x100ca1990 "walletTests", obj=<optimized out>) at ../../include/QtCore/../../src/corelib/kernel/qobjectdefs.h:409
#2355103 0x00007ffff7fa408c in QTest::qInvokeTestMethod(char const*, char const*) (slot=0x100ca1990 "walletTests") at qtestcase.cpp:1953
#2355104 0x00007ffff7fa408c in QTest::qInvokeTestMethod(char const*, char const*) (slotName=0x100948698 "walletTests()", data=data@entry=0x0) at qtestcase.cpp:2082
#2355105 0x00007ffff7fa4bb8 in QTest::qExec(QObject*, int, char**) (testObject=0x7fffffffd830) at qtestcase.cpp:2390
#2355106 0x00007ffff7fa4bb8 in QTest::qExec(QObject*, int, char**) (testObject=0x7fffffffd830, argc=<optimized out>, argv=<optimized out>) at qtestcase.cpp:2652
#2355107 0x000000010005d9df in main(int, char**) (argc=1, argv=0x7fffffffdd78) at qt/test/test_main.cpp:99
```

ryanofsky commented on src/qt/guiutil.h:
```diff
@@ -239,6 +240,82 @@ namespace GUIUtil
+        m_qstring = QString::fromStdString(std::move(str));
+    }
+};
+template<typename Q> AsStdStringWrapper<Q> AsStdString(Q& qstring) { return AsStdString<Q>(qstring); }
```
This should say `return AsStdStringWrapper<Q>(qstring)` or just `return {qstring}`

sipa replied:
Fixed! Thanks for figuring this out!

sipa commented:
Rebased.

sipa commented:
Removed the need for creating a vector with absolute values for serializing/deserializing BlockTransactionsRequest.

sipa commented:
Rebased.

sipa commented:
Rebased.

Christewart commented:
I ran this with #8469 -- which tests serialization code pretty extensively -- and it everything passed. 

If anyone wants to re-run they can do with this branch: https://github.com/Christewart/bitcoin/commits/pbt_noncastserial

sipa commented:
Rebased.

sipa committed:
Add Wrap helper

sipa commented:
I've reordered the commits again a bit, and made a small (but invasive) change to the design:
* Removed macro READWRITEAS (using `AsType<type>(obj)` instead)
* Added a `Wrap<wrapper>(obj)` function to later avoid the need for `COMPACTSIZE`, `VARINT`, and `LIMITED_STRING`, and removing the need for helper functions for each wrapper class.

Feel free to ignore the changes here; I'll keep taking small bits from it and submitting them as separate PRs.

sipa committed:
Convert wrapper

laanwj commented:
Concept ACK

sipa commented:
@laanwj I'm going to keep extracting smaller changes from this PR into separate ones (like #12916, #12886, #12752, #12740, #12712, #12683, #12658). No need to review the whole thing.

MarcoFalke commented:
Needs rebase due to merge of #12924

MarcoFalke commented:
Travis output:

```
A wild circular dependency in the form of "qt/guiutil -> qt/walletmodel -> qt/guiutil" appears...

jb55 commented:
Chris Stewart <notifications@github.com> writes:
> I ran this with #8469 -- which tests serialization code pretty extensively -- and it everything passed. 

I gave this a spin with the property tests as well. Seems to still pass.
There's a recent change to bench/prevector that needs to be updated but
other than that it appears to work.

sipa committed:
Convert prevector tests to new serialization

sipa commented:
Rebased.

practicalswift commented on src/compressor.h:
```diff
@@ -33,79 +29,94 @@ uint64_t DecompressAmount(uint64_t nAmount);
+        {
+            uint64_t v;
+            s >> VARINT(v);
+            m_val = DecompressAmount(v);
```
Make this conversion explicit since it changes signedness?

sipa replied:
The best I can do in this place is cast it to `I`, which isn't making anything interesting explicit, and will just hide the warning.

The issue is that CTxOut has a signed `int64_t`, but for UTXOs only ever contains unsigned values. I think the warning, if any, is appropriate, and should remain until the actual "issue" is resolved.

practicalswift replied:
Makes sense! Thanks for the clarification.

practicalswift commented on src/compressor.h:
```diff
@@ -33,79 +29,94 @@ uint64_t DecompressAmount(uint64_t nAmount);
+        explicit Wrapper(I& val) : m_val(val) {}
+        template<typename Stream> void Serialize(Stream& s) const
+        {
+            s << VARINT(CompressAmount(m_val));
```
Make this conversion explicit since it changes signedness?

practicalswift commented on src/compressor.h:
```diff
@@ -33,79 +29,94 @@ uint64_t DecompressAmount(uint64_t nAmount);
+    private:
+        O &txout;
+    public:
+        Wrapper(O &txoutIn) : txout(txoutIn) { }
```
Make this `explicit`? :-)

practicalswift commented on src/wallet/wallet.h:
```diff
@@ -583,17 +579,13 @@ class CWalletKey
-    inline void SerializationOp(Stream& s, Operation ser_action) {
+    SERIALIZE_METHODS(CWalletKey, obj)
+    {
         int nVersion = s.GetVersion();
```
The scope of `nVersion` can be reduced?

practicalswift commented on src/qt/walletmodel.h:
```diff
@@ -51,6 +51,65 @@ QT_BEGIN_NAMESPACE
+    private:
+        Q& m_qstring;
+    public:
+        Wrapper(Q& qstring) : m_qstring(qstring) {}
```
`explicit`?

practicalswift commented on src/test/dbwrapper_tests.cpp:
```diff
@@ -251,24 +251,26 @@ struct StringContentsSerializer {
-
-    template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action) {
-        if (ser_action.ForRead()) {
```
`ForRead` is no longer needed and can be removed from `serialize.h`?

practicalswift commented:
Concept ACK

@sipa has this been tested on a BE system?

laanwj commented:
This PR is really old by now, even though there was no disagreement. What is holding this up?

IMO. we should try to merge this after the 0.19 split-off in a few days.

practicalswift commented:
Has this been properly fuzz tested?

MarcoFalke commented:
@practicalswift Pretty much all of our fuzzers are  deserialization fuzzers

practicalswift commented:
@MarcoFalke Yes I know that :)  I was curious if these suggested serialization improvements had been subject to fuzz testing by the PR author - that is something different and hence my question :)

ryanofsky commented:
> @MarcoFalke Yes I know that :) I was curious if these suggested serialization improvements had been subject to fuzz testing by the PR author - that is something different and hence my question :)

I haven't paid much attention to fuzz testing, but I did spend a good amount of time reviewing this, and wonder if you could give more detail on the extra testing that would be useful here that goes beyond the normal testing.

It is probably safe to assume that if the author here went out of their way to do special testing, this would be mentioned in the description.

practicalswift commented:
> I haven't paid much attention to fuzz testing, but I did spend a good amount of time reviewing this, and wonder if you could give more detail on the extra testing that would be useful here that goes beyond the normal testing.

It would be useful to take this new code for a spin with our fuzzing test harnesses to make sure no new obvious issues are introduced.

This can be done using:

```
$ CC=clang CXX=clang++ ./configure --enable-fuzz --with-sanitizers=address,fuzzer,undefined
$ make
$ find src/test/fuzz/ -type f -executable
[…] then run the relevant fuzzers and observe any anomalies […]
```

> It is probably safe to assume that if the author here went out of their way to do special testing, this would be mentioned in the description.

TBH I'm not fuzz testing qualifies as special testing these days - in security critical project like ours it really should be standard procedure for this type of code :)

I've simply assumed that PR authors do fuzz testing routinely. I guess I'm mistaken :)

jb55 commented:
> @practicalswift Has this been properly fuzz tested?

both me and @Christewart ran the serialization property-based tests against this at one point. Would be good to revisit.

practicalswift commented:
@jb55 Great to hear! I'll try to break it by fuzzing once rebased :)

sipa commented:
I'll gladly rebase again next week, or try to cherry-pick pieces of it.

sipa commented:
Rebased.

sipa commented:
Rebased.

sipa commented:
Rebased.

sipa commented:
Rebased.

MarcoFalke commented:
Is this going to be split into smaller bites?

sipa commented:
I certainly can, though I think all the standalone improvements have already been split off and merged independently.

What remains is introducing new serialization macros, converting all serializers to use them, and then deleting the old ones. I'm happy to split those up into separate PRs if you think that's useful.

MarcoFalke commented:
Fair enough. I was asking for the general public

sipa commented:
Rebased on top of (a rebased version of) #17591, to test on a BE platform.

laanwj commented on src/qt/sendcoinsrecipient.h:
```diff
@@ -33,8 +57,8 @@ class SendCoinsRecipient
     QString message;
-    // Keep the payment request around as a serialized string to ensure
-    // load/store is lossless.
+    // If building with BIP70 is disabled, keep the payment request around as
```
This is no longer an "if" :smiley:

sipa replied:
Oops, accident revert added in a rebase. Fixed.

sipa commented:
Now also rebased on top of #17599 to get functional tests run on BE.

practicalswift commented:
Could bring in the new improved deserialization fuzz harness from #17225 to fuzz test this? The new fuzz harness does roundtrip testing where possible and covers more types. Would be nice to see this new code fuzz tested :)

---

Fuzzing in three easy steps:

```
$ CC=clang CXX=clang++ ./configure --enable-fuzz --with-sanitizers=address,fuzzer,undefined
$ make
$ find src/test/fuzz/ -type f -executable
[…] then run the relevant fuzzers and observe any anomalies […]
```

sipa commented:
Rebased, and fixes a bug in the deserialization fuzzing code.

practicalswift commented:
This needs to be addressed to make the sanitizers happy :)

```
blockencodings.h:83:34: runtime error: implicit conversion from type 'int' of 
    value 65536 (32-bit, signed) to type 'Differential::Wrapper<std::vector<unsigned short, std::allocator<unsigned short> > >::value_type'
    (aka 'unsigned short') changed the value to 0 (16-bit, unsigned)
```

laanwj commented:
code review ACK 9250a087d2f450f37342010aea0a5d583eea508b

sipa committed:
Switch to different wrapping model + different differential vector

sipa commented:
I've added a new commit with a few style improvements, and a new approach for writing wrapper classes, as well a different way of doing the differential encoding of compact block indexes.

Unless people object, I'll use these modified approaches in follow-up PRs that carve out pieces out of this PR.

ryanofsky commented:
Just reviewed commit "Switch to different wrapping model + different differential vector" (7da03b3ce5f4be94555b15bf4668bc251a8faac4) and I think it is a nice improvement. Getting rid of the various `Format::Wrapper` inner classes and pulling them into a standalone `Wrapper` class eliminates redundant code. And another really nice part of the change is replacing the `Differential` struct that had lots of operator overloads with simpler value function transform objects.

I did have some ideas and suggestions from looking at this, but feel free to take or leave them:

1. "Wrapper" terminology doesn't seem helpful when it isn't specifically referring to the actual wrapper class. Would maybe refer to the various classes that have `Ser` and `Unser` methods as formatters or helpers instead of wrappers since they don't really wrap anything. Similarly might rename `Wrap` function to `SerializeAs` or `FormatAs` and rename `W` template arguments to `F` or `H`.
2. I don't think there's a reason to require `Ser` and `Unser` methods to be static. It should be no problem to just call `W()::Ser()` instead of `W::Ser()` inside the new `Wrapper` class, and not requiring the methods to be static could be beneficial if the formatter/helper objects are used other contexts in the future. For example if a compression formatter was used to serialize many elements in a data structure, maybe it could save some state to avoid repeating work.

sipa commented:
@ryanofsky Thanks, I'll see how hard it is to address the non-static methods suggestion. What so you think of the name "Using" instead of "Wrap" ?

sipa commented:
Changed to terminology "formatters", with function "Using<Formatter>", and made the Ser/Unser functions (permitted to be) non-static.

ryanofsky commented:
> @ryanofsky Thanks, I'll see how hard it is to address the non-static methods suggestion. What so you think of the name "Using" instead of "Wrap" ?

I think it'd be good, assuming it's not too much trouble to rename. I also realized the names I suggested yesterday `SerializeAs` or `FormatAs` don't really make sense given how the template argument is used. More accurate would be `SerializeWith`, `FormatWith`, or `SerializeUsing`. No strong opinions about any of this, though.

laanwj commented:
Travis error:
```
blockencodings.h:68:17: runtime error: implicit conversion from type 'int' of value 65536 (32-bit, signed) to type 'unsigned short' changed the value to 0 (16-bit, unsigned)

    #0 0x561bb3939927 in DifferentialUnserTransform<unsigned short>::operator()(unsigned short) /home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-pc-linux-gnu/src/./blockencodings.h:68:17
    #1 0x561bb3938b15 in void VectorUsing<CompactSizeFormatter, DifferentialSerTransform<unsigned short>, DifferentialUnserTransform<unsigned short> >::Unser<CDataStream, std::vector<unsigned short, std::allocator<unsigned short> > >(CDataStream&, std::vector<unsigned short, std::allocator<unsigned short> >&) /home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-pc-linux-gnu/src/./serialize.h:625:29
    #2 0x561bb39387c3 in void Wrapper<VectorUsing<CompactSizeFormatter, DifferentialSerTransform<unsigned short>, DifferentialUnserTransform<unsigned short> >, std::vector<unsigned short, std::allocator<unsigned short> > >::Unserialize<CDataStream>(CDataStream&) /home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-pc-linux-gnu/src/./serialize.h:444:73
    #3 0x561bb39385bb in void Unserialize<CDataStream, Wrapper<VectorUsing<CompactSizeFormatter, DifferentialSerTransform<unsigned short>, DifferentialUnserTransform<unsigned short> >, std::vector<unsigned short, std::allocator<unsigned short> > >&>(CDataStream&, Wrapper<VectorUsing<CompactSizeFormatter, DifferentialSerTransform<unsigned short>, DifferentialUnserTransform<unsigned short> >, std::vector<unsigned short, std::allocator<unsigned short> > >&) /home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-pc-linux-gnu/src/./serialize.h:708:7
```

sipa commented:
Too many changes have been introduced compared to this PR in #17850, #17896, #17957, #18021, and #18112,  and I'm not going to rebase this on top of them. I'll keep picking changes from the branch, though, but no need to keep this open.
